src/client/dfs/dfs.c

```diff
@@ -6479,3 +6478,246 @@ dfs_obj_anchor_set(dfs_obj_t *obj, uint32_t index, daos_anchor_t *anchor)
+	if (rc) {
+		D_ERROR("daos_oit_open failed "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out_snap, rc = daos_der2errno(rc));
+	}
+
+	/** get and mark the SB and root OIDs */
+	d_iov_set(&marker, &mark_data, sizeof(mark_data));
+	rc = daos_oit_mark(oit, dfs->super_oid, &marker, NULL);
+	if (rc) {
+		D_ERROR("Failed to mark SB OID in OIT: "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out_oit, rc = daos_der2errno(rc));
+	}
+	rc = daos_oit_mark(oit, dfs->root.oid, &marker, NULL);
+	if (rc) {
+		D_ERROR("Failed to mark ROOT OID in OIT: "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out_oit, rc = daos_der2errno(rc));
+	}
+
+	/** iterate through the namespace and mark OITs starting from the root object */
+	while (!daos_anchor_is_eof(&anchor)) {
+		rc = dfs_iterate(dfs, &dfs->root, &anchor, &nr_entries, DFS_MAX_NAME * nr_entries,
+				 oit_mark_cb, (void *)oit.cookie);
+		if (rc) {
+			D_ERROR("dfs_iterate() failed: %d\n", rc);
+			D_GOTO(out_oit, rc);
+		}
+
+		nr_entries = DFS_NR_ITERATE;
+	}
+
+	/** Create lost+found directory and properly link unmarked oids there. */
+	if (flags & DFS_CHECK_LINK_LF) {
+		rc = dfs_open(dfs, NULL, "lost+found", S_IFDIR | 0755, O_CREAT | O_RDWR, 0, 0, NULL,
+			      &lf);
+		if (rc) {
+			D_ERROR("Failed to create lost+found directory: %d\n", rc);
+			D_GOTO(out_oit, rc);
+		}
+	}
+
+	/** list all unmarked oids */
+	memset(&anchor, 0, sizeof(anchor));
+	while (!daos_anchor_is_eof(&anchor)) {
+		nr_entries = DFS_NR_ITERATE;
+		rc = daos_oit_list_unmarked(oit, oids, &nr_entries, &anchor, NULL);
+		if (rc) {
+			D_ERROR("daos_oit_list_unmarked failed: "DF_RC"\n", DP_RC(rc));
+			D_GOTO(out_lf, rc = daos_der2errno(rc));
+		}
+
+		for (i = 0; i < nr_entries; i++) {
+			if (flags & DFS_CHECK_PRINT)
+				D_PRINT("oid["DF_U64"]: "DF_OID"\n", unmarked_entries,
+					DP_OID(oids[i]));
+
+			if (flags & DFS_CHECK_REMOVE) {
+				daos_handle_t oh;
+
+				rc = daos_obj_open(dfs->coh, oids[i], DAOS_OO_RW, &oh, NULL);
+				if (rc)
+					D_GOTO(out_oit, rc = daos_der2errno(rc));
+
+				rc = daos_obj_punch(oh, DAOS_TX_NONE, 0, NULL);
+				if (rc) {
+					daos_obj_close(oh, NULL);
+					D_GOTO(out_oit, rc = daos_der2errno(rc));
+				}
+
+				rc = daos_obj_close(oh, NULL);
+				if (rc)
+					D_GOTO(out_oit, rc = daos_der2errno(rc));
+			}
+
+			if (flags & DFS_CHECK_LINK_LF) {
+				struct dfs_entry	entry = {0};
+				enum daos_otype_t	otype = daos_obj_id2type(oids[i]);
+				struct timespec		now;
+				char			name[DFS_MAX_NAME + 1];
+				daos_size_t		len;
+
+				entry.uid = geteuid();
+				entry.gid = getegid();
+				oid_cp(&entry.oid, oids[i]);
+				if (daos_is_array_type(otype))
+					entry.mode = S_IFREG | 0755;
+				else
+					entry.mode = S_IFDIR | 0755;
+				rc = clock_gettime(CLOCK_REALTIME, &now);
+				if (rc)
+					D_GOTO(out_lf, rc = errno);
+				entry.atime = entry.mtime = entry.ctime = now.tv_sec;
+				entry.atime_nano = entry.mtime_nano = entry.ctime_nano =
+					now.tv_nsec;
+				entry.chunk_size = dfs->attr.da_chunk_size;
```

Comment:
I think it would be easier to call this once before the loop as well and just have all recovered files having the same timestamp.

Reply:
sure